src/main/GCNodeAllocator.cpp

```diff
@@ -0,0 +1,511 @@
+      updateHeapBounds(result, actual_bytes);
+    }
+  }
+  if (!result) {
+    allocerr("failed to allocate object");
+  }
+
+#ifdef ALLOCATION_CHECK
+  if (lookup_in_allocation_map(result)) {
+    allocerr("reusing live allocation");
+  }
+  add_to_allocation_map(result, actual_bytes);
+  lookupPointer(result);  // Check lookup table consistency.
+#endif
+
+#ifdef HAVE_ADDRESS_SANITIZER
+  // Poison the redzones.
+  // The end redzone can be larger than s_redzone_size if the allocator
+  // grew the allocation larger than the requested size.
+  // The additional size (actual_bytes - bytes) is added to the default
+  // redzone size.
+  void* end_redzone = offsetPointer(result, bytes - s_redzone_size);
+  unsigned end_redzone_size = s_redzone_size + (actual_bytes - bytes);
+  ASAN_POISON_MEMORY_REGION(result, s_redzone_size);
+  ASAN_POISON_MEMORY_REGION(end_redzone, end_redzone_size);
+
+  // Offset the result pointer past first redzone.
+  result = offsetPointer(result, s_redzone_size);
+#endif // HAVE_ADDRESS_SANITIZER
+  return result;
+}
+
+void rho::GCNodeAllocator::free(void* pointer) {
+#ifdef HAVE_ADDRESS_SANITIZER
+  // Adjust for redzone to find the true start of the allocation.
+  pointer = offsetPointer(pointer, -s_redzone_size);
+
+  // Unpoison the first redzone so we can link it into a freelist.
+  ASAN_UNPOISON_MEMORY_REGION(pointer, sizeof(FreeListNode));
+#endif
+#ifdef ALLOCATION_CHECK
+  if (!lookupPointer(pointer)) {
+    allocerr("can not free unknown/already-freed pointer");
+  }
+  remove_from_allocation_map(pointer);
+#endif
+  uintptr_t pointer_uint = reinterpret_cast<uintptr_t>(pointer);
+  AllocatorSuperblock* superblock =
+      AllocatorSuperblock::arenaSuperblockFromPointer(pointer_uint);
+  if (superblock) {
+    superblock->freeBlock(pointer);
+  } else {
+    // Search for the allocation in the hashtable.  If the allocation is in a
+    // superblock, the superblock is left as is but the superblock bitset is
+    // updated to indicate that the block is free.
+    // If the allocation is a large object (ie. not in a superblock), then
+    // the object is taken out of the hashtable and inserted into a freelist.
+    AllocationTable::Allocation* allocation =
+        s_alloctable->search(pointer_uint);
+    if (!allocation) {
+      // Free failed: could not find block to free.
+      allocerr("failed to free pointer - unallocated or double-free problem");
+    }
+    if (allocation->isSuperblock()) {
+      AllocatorSuperblock* superblock = allocation->asSuperblock();
+      superblock->freeBlock(pointer);
+      // Done. Don't erase hash entries for the superblock.
+    } else if (allocation->asPointer() == pointer) {
+      // Erase all entries in hashtable for the allocation.
+      s_alloctable->erase(pointer_uint, allocation->sizeLog2());
+      // Add allocation to free list.
+      addLargeAllocationToFreelist(pointer, allocation->sizeClass());
```

Comment:
A maximum limit for the number of active freelist nodes could be added and then when it overflows the oldest freelist entries could be released.

Alternatively it would be possible to have a limit on the number of free byte in all freelists, and when the limit is reached all free nodes could be released.